File: Data Structures and Algorithms/Python/Stack.py
#
# Ayush Kumar (202103103510253)
# 4th Sem Div: A
# Stack.py
#

class Stack:
  __stack = []

  def __init__(self, size):
    if(size <= 0):
      raise Exception("Invalid Size")
      return
    self.__size = size

  def push(self, value):
    if (len(self.__stack) >= self.__size):
      print("Stack overflow")
      return
    self.__stack.append(value)

  def pop(self):
    if (len(self.__stack) == 0):
      print("Stack underflow")
      return
    self.__stack.pop(-1)

  def peep(self, index=None):
    if (index == None):
      return self.__stack[-1]

    if (index >= len(self.__stack) or index < 0):
      print("Stack out of index")
      return
    return self.__stack[index]

  def getTop(self):
    return len(self.__stack) - 1

  def change(self, index, value):
    if (index >= len(self.__stack) or index < 0):
      print("Stack out of index")
      return
    self.stack[index] = value

  def display(self):
    print(self.__stack)

File: Data Structures and Algorithms/C++/Stack.cpp
/*
    Ayush Kumar (202103103510253)
    4th Sem Div: A
    Stack.cpp
*/

#include "Stack.hpp"
#include <iostream>

using namespace std;

Stack::Stack(int size) {
  if (size <= 0) {
    throw invalid_argument("size");
    return;
  }

  this->stack = new int[size];
  this->size = size;
}

void Stack::change(int index, int value) {
  if (this->top < index || index < 0) {
    cout << "Stack Index out of bound" << endl;
    return;
  }

  this->stack[index] = value;
}

void Stack::pop() {
  if (this->top <= -1) {
    cout << "Stack Underflow" << endl;
  }

  this->stack[this->top--] = NULL;
}

int Stack::peep(int index) {
  if (index > this->top) {
    cout << "Stack Index out of bound" << endl;
    return -1;
  }

  return this->stack[index];
}

void Stack::display() {
  cout << "{ ";
  for (int i = 0; i <= this->top; i++) {
    cout << this->stack[i] << " ";
  }
  cout << "}" << endl;
}

void Stack::push(int value) {
  if (this->top >= this->size) {
    cout << "Stack Overflow" << endl;
    return;
  }

  this->stack[++this->top] = value;
}

int Stack::peep() { return Stack::peep(this->top); }

int Stack::getTop() { return this->top; }